README.md
# OMAKOView

[![Version](https://img.shields.io/cocoapods/v/OMAKOView.svg?style=flat)](http://cocoapods.org/pods/OMAKOView)
[![License](https://img.shields.io/cocoapods/l/OMAKOView.svg?style=flat)](http://cocoapods.org/pods/OMAKOView)
[![Platform](https://img.shields.io/cocoapods/p/OMAKOView.svg?style=flat)](http://cocoapods.org/pods/OMAKOView)
[![Readme Score](http://readme-score-api.herokuapp.com/score.svg?url=https://github.com/kgleong/omakoview)](http://clayallsopp.github.io/readme-score?url=https://github.com/kgleong/omakoview)
[![CI Status](http://img.shields.io/travis/kgleong/OMAKOView.svg?style=flat)](https://travis-ci.org/kgleong/OMAKOView)

|Description|Specification|
|----|-----------|
|Language| Swift 3|

## Overview

Custom views included in this library:

|Name|Description|
|----|-----------|
|[Pop Up View & Loading Spinner](#pop-up-view)|A pop up that shows a title and/or caption in addition to a loading spinner.|
|[Star View](#star-view)|Highly configurable `UIView` star drawn using `UIBezierPath` objects.|
|[Partially visible swipeable view](#partially-visible-swipeable-view)|A partially visible view that responds to vertical swipes to reveal and hide its content.|

### Table of Contents

* [PopUpView](#pop-up-view)
* [Star View](#star-view)
* [Partially visible swipeable view](#partially-visible-swipeable-view)

## Pop Up View

Customizable pop up that can fade away after a specified interval and also takes a closure that is called on dismissal or completion.

![Pop up view demo](/images/popup-view-demo.gif)

### Configuration

|Property Name|Description|Type|Default|
|-------------|-----------|----|-------|
|`titleText`|Styleable title text|`NSMutableAttributedString`|N/A|
|`bodyText`|Styleable body text|`NSMutableAttributedString`|N/A|
|`titleFontName`|Title font name. Passed into `UIFont(name:size:)`|`String`|`nil`.  Will default to system font.|
|`bodyFontName`|Body font name. Passed into `UIFont(name:size:)`|`String`|`nil`.  Will default to system font.|
|`titleFontColor`|Tile font color|`UIColor`|Dark blue|
|`bodyFontColor`|Body font color|`UIColor`|Purple|
|`spinnerBorderColor`|Spinner border color|`UIColor`|Off white|
|`spinnerFillColor`|Used for spinners which retain the same color.|`UIColor`|Powder blue|
|`spinnerBorderWidth`|The border width of the spinner.|`CGFloat`|5.0|
|`padding`|The padding between the elements.  E.g., spinner, title, body.|`CGFloat`|10.0|
|`cornerRadius`|If set, will clip the view and round the edges to the specified value.|`CGFloat`|10.0|
|`spinnerSizeInPoints`|The size of the spinner view.|`CGFloat`|20.0|
|`spinnerDuration`|The duration of the rotation and any other spinner animations.|`CGFloast`|2.0|

### Usage

#### Text only

Display a standard pop up with a title and body.

```swift
popUpView = OMAKOPopUpView()

popUpView.titleText = NSMutableAttributedString(string: "Title")
popUpView.bodyText = NSMutableAttributedString(string: "Body")

// Pass in the view that the popup should be centered in.
// The pop up will add itself to the view's hierarchy.
popUpView.display(parentView: view)
```

Hiding a popup:
```swift
popUpView.hide(completion: nil)

// Or a completion block can be passed in:
popUpView.hide() { print("Do something after the popup is dismissed") }
```

Displaying a popup for a set interval:

```swift
/*
    Will dismiss the popup after 5 seconds.

    A completion block can be passed in, which will execute after the popup
    is dismissed.
*/
popUpView.display(parentView: view, withDuration: 5.0, completion: nil)
```

#### Spinners

```swift
popUpView = OMAKOPopUpView()

popUpView.titleText = NSMutableAttributedString(string: "Loading")
popUpView.spinnerSizeInPoints = 40.0 // Results in a larger spinner.

popUpView.displaySpinner(parentView: view, spinnerType: .square) // Square spinner
popUpView.displaySpinner(parentView: view, spinnerType: .star) // Star spinner

// Hide
popUpView.hide(completion: nil)
```

## Star View

A `UIView` that draws a highly configurable star using `UIBezierPath` objects.

![Star view demo](/images/star-view-demo.gif)

### Configuration

In Interface Builder, the following properties are available:

![Star view inspectable properties](/images/star-inspectable.png)

|Property Name|Description|Type|Default|
|-------------|-----------|----|-------|
|`strokeWidth`|The star's border width.|`CGFloat`|`5.0`|
|`hasStroke`|If true, a border will be rendered.|`Bool`|`false`|
|`strokeColor`|Border color.|`UIColor`|Red|
|`fillColor`|The star's fill color.|`UIColor`|Powder blue|
|`innerToOuterRadiusRatio`|Determines the thickness of the star.  Values closer to 1.0 will widen the arms of the star, while values closer to 0.0 will thin the arms.|`CGFloat`|`0.45`|
|`starToViewRatio`|The size of the star proportional to the view.  Range: 0.0 to 1.0.|`CGFloat`|`1.0`|
|`cacheVertices`|Vertex positions will be cached.  Set this to false if vertex positions are expected to change between `draw(_:)` render calls.|`Bool`|`true`|

### Usage

![Star view variations](/images/star_variations.png)

```swift
var starView = OMAKOStarView()
starView.hasStroke = true
starView.strokeWidth = 5.0
starView.strokeColor = UIColor.red
starView.fillColor = UIColor.yellow
starView.innerToOuterRadiusRatio = 0.25 // Results in a thinner star.
starView.starToViewRatio = 0.8 // Results in a smaller star.
```

Rounding the edges inscribes the star within a circle, since the view is a subclass of `UIView`.

```swift
var starView = OMAKOStarView()

// Creates a circular border. The usual UIView properties are available.
starView.layer.cornerRadius = starView.bounds.width/Float(2)
starView.layer.borderWidth = 5.0
starView.layer.borderColor = UIColor.red.cgColor
starView.clipToBound = true
```

## Partially Visible Swipeable View

![Partially Visible Swipeable View demo](/images/partially-visible-swipeable-demo-1.gif)

### Usage

#### Important Note on Constraints

**A width constraint for the view must be set if using Interface Builder.**

The `remove at build time` placeholder checkbox **MUST BE** selected for the width constraint.

Interface builder autogenerates constraints, and this will result in constraint conflicts if the above steps are not followed.

The same needs to be done for **height** if the height of the view is dynamic.  E.g., multiline labels that contain dynamic text.

#### `UIViewController` Example

```swift
class ViewController: UIViewController {
    @IBOutlet weak var containerView: OMAKOPartiallyVisibleSwipeableView!
    @IBOutlet weak var titleLabel: UILabel!

    let titleLoremIpsum = "Qui officia deserunt anim id est laborum."

    override func viewDidLoad() {
        super.viewDidLoad()

        // Make labels multiline and add text
        titleLabel.numberOfLines = 0
        titleLabel.text = titleLoremIpsum

        /*
            `setupView(bottomLayoutGuide:)` should be called after any
            changes are made to subviews that may affect the size of the
            `containerView`.  E.g., Dynamically setting multiline label text.

            Optionally, `setupView()` can be called without paramters and this
            will by default pin the view to the bottom of the superview.

           Pinning the view to the `bottomLayoutGuide`, however, is preferred.
        */
        containerView.setupView(bottomLayoutGuide: bottomLayoutGuide)
    }

    override func didRotateFromInterfaceOrientation(fromInterfaceOrientation: UIInterfaceOrientation) {
        // Must be called on device orientation changes.
        // Screen rotations will change the height of the `container view`,
        // and `onRotate()` accounts for any changes.
        containerView.onRotate()
    }
}
```

### IBInspectable Properties

![Partially Visible Swipeable IBInspectable properties](/images/partially-visible-swipeable-ib-inspectable.png)

|Name|Type|Description|Values|
|----|----|-----------|------|
|**visible amount**|`CGFloat`|Proportion of the view's height that should be initially visible.|A value of `1.0` is fully visible, and `0.0` is completely hidden. Default value is `0.4`, which makes 40% of the view's height visible.|
|**relative width**|`CGFloat`|Width of the view relative to its superview.|`1.0` matches the width of the parent, and `0.0` gives the view a width of 0.  Default is `0.75` or three-quarters of the superview's width |
|**align center**|`Boolean`|Centers the view within the superview.|Default set to `true`.|

The properties below adjust the animation oscillation and bounce .  See the Apple developer documentation on [`animateWithDuration ` with spring damping](https://developer.apple.com/reference/uikit/uiview/1622594-animatewithduration?language=objc) for more details.

|Name|Type|Description|Values|
|----|----|-----------|------|
|**duration**|`CGFloat`|How long the animation should take, in seconds.|If `<= 0`, no animation will be applied.  Default is set to `1.5` seconds|
|**damping ratio**|`CGFloat`| Amount of bouncing desired.|Use `1` to smoothly decelerate with no oscillation.  Values closer to `0` will increase oscillation/bouncing. The higher the value, the faster the oscillations will cease. Default is set to `0.5`.|
|**spring speed**|`CGFloat`| The spring's initial velocity. The damping ratio will proportionally decrement this value until the velocity reaches 0.|Higher values will increase the oscillations' speed. Default is set to `0.5`.|

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Installation

OMAKOView is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "OMAKOView"
```

## Author

Kevin Leong, kgleong@gmail.com

## License

OMAKOView is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.2.1] - 2017-06-12

### Changed

* Replaces `M_PI` with `Double.pi`.

## [2.2.0] - 2016-12-10

### Changed

* Added `OMAKOPopUpView`.
* Added `OMAKOStarView`.

## [2.1.3] - 2016-11-25

### Changed

* Updated library and example project to Swift 3.


OMAKOView/Classes/OMAKOPopUpView.swift
import UIKit

public enum OMAKOSpinnerType {
    case square
    case star
}

public enum OMAKOPopUpViewError: Error {
    case missingContent
}

open class OMAKOPopUpView: UIView {

    // MARK: - Public variables

    /// Text

    open var titleText: NSMutableAttributedString?
    open var bodyText: NSMutableAttributedString?

    open var titleFontName: String?
    open var bodyFontName: String?

    open var titleFontSize: CGFloat = 17.0
    open var bodyFontSize: CGFloat = 15.0

    open var titleFontColor: UIColor = UIColor(red: 0.1, green: 0.1, blue: 0.3, alpha: 1.0)
    open var bodyFontColor: UIColor = UIColor(red: 0.3, green: 0.1, blue: 0.3, alpha: 1.0)

    /// Spinner options
    open var spinnerBorderColor: UIColor = UIColor(red: 0.9, green: 0.9, blue: 0.9, alpha: 1.0)
    open var spinnerFillColor: UIColor = UIColor(red: 0.05, green: 0.5, blue: 1.0, alpha: 1.0)
    open var spinnerBorderWidth: CGFloat = 5.0

    /// Layout

    open var padding: CGFloat = 10.0
    open var cornerRadius: CGFloat? = 10.0
    open var spinnerSizeInPoints: CGFloat = 20.0
    open var spinnerDuration: TimeInterval = 2.0

    // MARK: - Private variables

    fileprivate var spinnerView: UIView?
    fileprivate var titleLabel: UILabel?
    fileprivate var bodyLabel: UILabel?

    fileprivate var defaultBackgroundColor = UIColor(red: 0.25, green: 0.25, blue: 0.40, alpha: 0.5)

    fileprivate var spinnerType: OMAKOSpinnerType?

    fileprivate var constraintList = [NSLayoutConstraint]()

    // MARK: - UIView Methods

    override open func didMoveToSuperview() {
        super.didMoveToSuperview()

        /*
         `didMoveToSuperview` and `willMoveToSuperview` are called
         by `removeFromSuperview`, which calls `willMove(toSuperview: nil)`.

         `nil` check prevents re-creating constraints and views when
         the pop up is removed from its parent view.
         */
        if superview != nil {
            setupView()

            // TODO: Ideally, a check should be performed earlier, (e.g., in the public API methods) 
            // rather than after the view has been added to the view hierarchy.
            do {
                try setupConstraints()
            }
            catch OMAKOPopUpViewError.missingContent {
                removeFromSuperview()
                print("No content provided.  Removing from superview.")
            }
            catch {
                print("Error occurred: \(error)")
            }
        }
    }

    // MARK: - Public API

    open func display(parentView: UIView) {
        display(parentView: parentView, withDuration: nil, completion: nil)
    }

    open func display(parentView: UIView, withDuration duration: TimeInterval?, completion: (() -> Void)?) {
        parentView.addSubview(self)
        parentView.bringSubview(toFront: self)

        guard let duration = duration else {
            completion?()
            return
        }

        UIView.animate(withDuration: duration, animations: {
            self.alpha = 0.0
        }, completion: { (Boolean) -> Void
            in
            
            self.removeFromSuperview()

            guard let completion = completion else {
                return
            }

            completion()
        })
    }

    open func displaySpinner(parentView: UIView, spinnerType: OMAKOSpinnerType) {
        self.spinnerType = spinnerType

        parentView.addSubview(self)
        animateSpinner()
    }

    open func hide(completion: (() -> Void)?) {
        guard superview != nil else {
            print("Pop up must be part of a view hierarchy.  No superview detected.")
            return
        }

        removeFromSuperview()
        completion?()
    }

    // MARK: - Spinner Animations

    fileprivate func animateSpinner() {
        guard let spinnerType = spinnerType else {
            return
        }

        switch spinnerType {
        case .square:
            animateSquareSpinner()
        case .star:
            animateStarSpinner()
        }
    }

    fileprivate func animateStarSpinner() {
        rotateSpinner()
    }

    fileprivate func animateSquareSpinner() {
        guard let spinnerView = spinnerView, let startingColor = spinnerView.backgroundColor else {
            return
        }

        /// Continuous cycling through colors
        UIView.animateKeyframes(
            withDuration: spinnerDuration,
            delay: 0,
            options: [UIViewKeyframeAnimationOptions.repeat],
            animations: {
                guard let spinnerView = self.spinnerView else {
                    return
                }

                UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 1.0/3.0) {
                    spinnerView.backgroundColor = UIColor.green
                }
                UIView.addKeyframe(withRelativeStartTime: 1.0/3.0, relativeDuration: 1.0/3.0) {
                    spinnerView.backgroundColor = UIColor.blue
                }
                UIView.addKeyframe(withRelativeStartTime: 2.0/3.0, relativeDuration: 1.0/3.0) {
                    spinnerView.backgroundColor = startingColor
                }
            },
            completion: nil
        )

        rotateSpinner()
    }

    fileprivate func rotateSpinner() {
        guard let spinnerView = spinnerView else {
            return
        }

        /// Continuous rotation
        let animation = CABasicAnimation(keyPath: "transform.rotation")
        animation.fromValue = 0
        animation.toValue = CGFloat(Double.pi * 2)
        animation.duration = spinnerDuration
        animation.repeatCount = Float.infinity

        spinnerView.layer.add(animation, forKey: nil)
    }

    // MARK: - Constraint Setup

    fileprivate func setupConstraints() throws {
        try setupSelfConstraints()
        setupSpinnerViewConstraints()

        NSLayoutConstraint.activate(constraintList)
    }

    fileprivate func setupSelfConstraints() throws {
        /// Horizontal and vertical centering

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .centerX,
                relatedBy: .equal,
                toItem: superview,
                attribute: .centerX,
                multiplier: 1,
                constant: 0)
        )

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .centerY,
                relatedBy: .equal,
                toItem: superview,
                attribute: .centerY,
                multiplier: 1,
                constant: 0)
        )

        /// Width

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .width,
                relatedBy: .greaterThanOrEqual,
                toItem: titleLabel,
                attribute: .width,
                multiplier: 1,
                constant: padding * 2
            )
        )

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .width,
                relatedBy: .greaterThanOrEqual,
                toItem: bodyLabel,
                attribute: .width,
                multiplier: 1,
                constant: padding * 2
            )
        )

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .width,
                relatedBy: .greaterThanOrEqual,
                toItem: spinnerView,
                attribute: .width,
                multiplier: 1,
                constant: padding * 2
            )
        )

        constraintList.append(
            NSLayoutConstraint(
                item: self,
                attribute: .width,
                relatedBy: .lessThanOrEqual,
                toItem: superview,
                attribute: .width,
                multiplier: 0.5,
                constant: 0
            )
        )

        centerHorizontallyInParent(subview: titleLabel)
        centerHorizontallyInParent(subview: bodyLabel)

        try setupPaddingConstraints()
    }

    fileprivate func setupPaddingConstraints() throws {
        let orderedSubviews = try orderSubviews()

        if let firstViewInPopUp = orderedSubviews.first {
            var topPadding: CGFloat = padding

            /// Spinners require more top padding due to
            /// layer rotation.
            if firstViewInPopUp == spinnerView {
                topPadding = padding * 1.3
            }

            constraintList.append(
                NSLayoutConstraint(
                    item: firstViewInPopUp,
                    attribute: .top,
                    relatedBy: .equal,
                    toItem: self,
                    attribute: .top,
                    multiplier: 1,
                    constant: topPadding
                )
            )
        }

        if let lastViewInPopUp = orderedSubviews.last {
            constraintList.append(
                NSLayoutConstraint(
                    item: self,
                    attribute: .bottom,
                    relatedBy: .equal,
                    toItem: lastViewInPopUp,
                    attribute: .bottom,
                    multiplier: 1,
                    constant: padding
                )
            )
        }

        /// Padding between subviews
        for (index, subview) in orderedSubviews.enumerated() {
            if(index > 0) {
                let previousSubview = orderedSubviews[index - 1]

                constraintList.append(
                    NSLayoutConstraint(
                        item: subview,
                        attribute: .top,
                        relatedBy: .equal,
                        toItem: previousSubview,
                        attribute: .bottom,
                        multiplier: 1,
                        constant: padding
                    )
                )
            }
        }
    }

    fileprivate func orderSubviews() throws -> [UIView] {
        var orderedSubviews = [UIView]()

        if let spinnerView = spinnerView {
            orderedSubviews.append(spinnerView)
        }

        if let titleLabel = titleLabel {
            orderedSubviews.append(titleLabel)
        }

        if let bodyLabel = bodyLabel {
            orderedSubviews.append(bodyLabel)
        }

        guard !orderedSubviews.isEmpty else {
            throw OMAKOPopUpViewError.missingContent
        }

        return orderedSubviews
    }

    fileprivate func setupSpinnerViewConstraints() {
        guard let spinnerView = spinnerView else {
            return
        }

        centerHorizontallyInParent(subview: spinnerView)

        /// Width and Height

        constraintList.append(
            NSLayoutConstraint(
                item: spinnerView,
                attribute: .width,
                relatedBy: .equal,
                toItem: nil,
                attribute: .notAnAttribute,
                multiplier: 1,
                constant: spinnerSizeInPoints
            )
        )

        constraintList.append(
            NSLayoutConstraint(
                item: spinnerView,
                attribute: .height,
                relatedBy: .equal,
                toItem: nil,
                attribute: .notAnAttribute,
                multiplier: 1,
                constant: spinnerSizeInPoints
            )
        )
    }

    fileprivate func centerHorizontallyInParent(subview: UIView?) {
        guard let subview = subview else {
            return
        }

        constraintList.append(
            NSLayoutConstraint(
                item: subview,
                attribute: .centerX,
                relatedBy: .equal,
                toItem: self,
                attribute: .centerX,
                multiplier: 1,
                constant: 0
            )
        )
    }

    // MARK: - View Setup

    fileprivate func setupView() {
        setupSelfView()
        setupSpinnerView()
        setupTitleLabel()
        setupBodyLabel()
    }

    fileprivate func setupSelfView() {
        translatesAutoresizingMaskIntoConstraints = false

        if backgroundColor == nil {
            backgroundColor = defaultBackgroundColor
        }

        if let cornerRadius = cornerRadius {
            clipsToBounds = true
            layer.cornerRadius = cornerRadius
        }
    }

    fileprivate func setupSpinnerView() {
        guard let spinnerType = spinnerType else {
            return
        }

        switch spinnerType {
        case .square:
            setupSquareSpinner()
        case .star:
            setupStarSpinner()
        }

        guard let spinnerView = spinnerView else {
            return
        }

        spinnerView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(spinnerView)
    }

    fileprivate func setupStarSpinner() {
        let starView = OMAKOStarView()

        starView.innerToOuterRadiusRatio = 0.45
        starView.starToViewRatio = 1.0
        starView.fillColor = spinnerFillColor
        starView.strokeColor = spinnerBorderColor
        starView.strokeWidth = spinnerBorderWidth
        starView.hasStroke = true
        starView.backgroundColor = UIColor.clear

        spinnerView = starView
    }

    fileprivate func setupSquareSpinner() {
        let squareView = UIView()

        squareView.backgroundColor = UIColor.red
        squareView.layer.cornerRadius = spinnerSizeInPoints / 4
        squareView.layer.borderWidth = 2
        squareView.layer.borderColor = spinnerBorderColor.cgColor

        spinnerView = squareView
    }

    fileprivate func setupTitleLabel() {
        titleLabel = createLabel(
            attributedText: titleText,
            fontName: titleFontName,
            fontSize: titleFontSize,
            fontColor: titleFontColor,
            boldByDefault: true
        )

        guard let titleLabel = titleLabel else {
            return
        }

        titleLabel.numberOfLines = 0
        titleLabel.translatesAutoresizingMaskIntoConstraints = false

        addSubview(titleLabel)
    }

    fileprivate func setupBodyLabel() {
        bodyLabel = createLabel(
            attributedText: bodyText,
            fontName: bodyFontName,
            fontSize: bodyFontSize,
            fontColor: bodyFontColor,
            boldByDefault: false
        )

        guard let bodyLabel = bodyLabel else {
            return
        }

        bodyLabel.numberOfLines = 0
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false

        addSubview(bodyLabel)
    }

    fileprivate func createLabel(
            attributedText: NSMutableAttributedString?,
            fontName: String?,
            fontSize: CGFloat,
            fontColor: UIColor,
            boldByDefault: Bool = false
        ) -> UILabel? {

        guard let attributedText = attributedText else {
            return nil
        }

        var font: UIFont?

        if let fontName = fontName {
            font = UIFont(name: fontName, size: fontSize)
        }
        else if boldByDefault {
            font = UIFont.boldSystemFont(ofSize: fontSize)
        }
        else {
            font = UIFont.systemFont(ofSize: fontSize)
        }

        let fullRange: NSRange = NSMakeRange(0, attributedText.length)

        attributedText.addAttribute(NSFontAttributeName, value: font!, range: fullRange)
        attributedText.addAttribute(NSParagraphStyleAttributeName, value: defaultParagraphStyle(), range: fullRange)
        attributedText.addAttribute(NSForegroundColorAttributeName, value: fontColor, range: fullRange)

        let label = UILabel()
        label.translatesAutoresizingMaskIntoConstraints = false
        label.attributedText = attributedText

        return label
    }

    fileprivate func defaultParagraphStyle() -> NSMutableParagraphStyle {
        let paragraphStyle: NSMutableParagraphStyle = NSMutableParagraphStyle()
        paragraphStyle.alignment = .center

        return paragraphStyle
    }
}
